bobtimus/src/loan.rs

```diff
@@ -44,42 +45,520 @@ pub struct LoanOffer {
     /// The max_ltv protects the lender from Bitcoin falling too much.
     pub max_ltv: Decimal,
 
-    pub interest: Vec<Interest>,
+    /// Base interest in percent (to be applied to the principal amount)
+    pub base_interest_rate: Decimal,
+
+    /// Interest in relation to terms
+    pub terms: Vec<Term>,
+
+    /// Interest rates in relation to collteralization
+    pub collateralizations: Vec<Collateralization>,
 }
 
-#[derive(Debug, Clone, serde::Serialize)]
-pub struct Interest {
-    /// Loan term in days
-    pub term: u32,
-    /// Collateralization in percent
-    ///
-    /// Rational: If a borrower over-collateralizes with e.g. 150% -> better rate than at 140%
+#[derive(Debug, Clone, Copy, serde::Serialize)]
+pub struct Term {
+    pub days: u32,
+    /// Interest to be added on top of the base interest rate for this term
+    pub interest_mod: Decimal,
+}
+
+/// Allows to specify a better rate for users that
+#[derive(Debug, Clone, Copy, serde::Serialize)]
+pub struct Collateralization {
     pub collateralization: Decimal,
-    /// Interest rate in percent
-    pub interest_rate: Decimal,
+    /// Interest to be added on top of the base interest rate for this term.
+    pub interest_mod: Decimal,
 }
 
+// TODO: Make sure that removing sat_per_vbyte is OK here
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
 pub struct LoanRequest {
-    #[serde(with = "::elements::bitcoin::util::amount::serde::as_sat")]
-    pub collateral_amount: Amount,
-    collateral_inputs: Vec<Input>,
-    #[serde(with = "::elements::bitcoin::util::amount::serde::as_sat")]
-    fee_sats_per_vbyte: Amount,
-    borrower_pk: PublicKey,
+    pub principal_amount: LiquidUsdt,
+    pub collateral_amount: LiquidBtc,
+    pub collateral_inputs: Vec<Input>,
+    pub collateralization: Decimal,
+    pub borrower_pk: PublicKey,
     /// Loan term in days
     pub term: u32,
-    borrower_address: Address,
+    pub borrower_address: Address,
+}
+
+pub fn calculate_interest_rate(
+    borrower_term: u32,
+    borrower_collateralization: Decimal,
+    term_thresholds: &[Term],
+    collateralization_thresholds: &[Collateralization],
+    base_interest_rate: Decimal,
+) -> Result<Decimal> {
+    let mut term_interest_mod = Decimal::ZERO;
+    for term in term_thresholds {
+        if borrower_term >= term.days {
+            term_interest_mod = term.interest_mod;
+            continue;
+        }
+        break;
+    }
+
+    let mut collateralization_interest_mod = Decimal::ZERO;
+    for collateralization in collateralization_thresholds {
+        if borrower_collateralization >= collateralization.collateralization {
+            collateralization_interest_mod = collateralization.interest_mod;
+            continue;
+        }
+        break;
+    }
```

Comment:
I don't understand the logic here :) Can you explain this please?

Reply:
Overall idea: The lender can define different interest rates based on term and collateralization. The `LoanOffer` contains term and collateralization thresholds that may modify the interest rate of the loan. For collateralizations the assumption is that the modifier would usually be `-x` (if higher collateraliztion then lower interest rate), for the term it would be `+x` (if longer term then higher interest rate).  

This piece of code: The modifier is set inside the if. As long as the collateralization falls into a higher threshold we keep looping. If the threshold is not higher we stop and use the modifier that was last set.

Note that for the term this has to be more restrictive - I will add a check to the loan acceptance criteria for that.

Reply:
I feel dumb and don't understand it 😔